### PR TITLE
Add HDFSArchiver tool

### PR DESCRIPTION
**Pull request by ebysans**
1. Add HDFSArchiver class, which is a tool originally in the refinery-core module of wikimedia analytics refinery source repo.
2. Add the test file for HDFSArchiver to test folder.

Bug: T316120

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -127,6 +127,12 @@
             <version>3.7.1</version>
         </dependency>
 
+        <dependency>
+            <groupId>org.wikimedia.analytics.refinery.tools</groupId>
+            <artifactId>refinery-tools</artifactId>
+            <version>0.2.4</version>
```
Comment (jobar) [inline pom.xml]: This should say `0.2.5` :)

File: src/test/scala/org/wikimedia/analytics/hdfstools/TestHDFSArchiver.scala [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        }
+    }
+
+    def setupSource(bkpDir: String, sourceDir: String): Unit = {
+        Files.createDirectory(Paths.get(sourceDir))
+        //    Files.copy(Paths.get(bkpDir), Paths.get(sourceDir))
+        Files.copy(Paths.get(bkpDir, "_SUCCESS"), Paths.get(sourceDir, "_SUCCESS"))
+        Files.copy(Paths.get(bkpDir, "archive.gz"), Paths.get(sourceDir, "archive.gz"))
+    }
+}
```
Comment (jobar) [inline src/test/scala/org/wikimedia/analytics/hdfstools/TestHDFSArchiver.scala]: There should a newline at the end of this file

File: src/main/scala/org/wikimedia/analytics/hdfstools/HDFSArchiver.scala [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        fs.delete(archiveFile, false)
+        val result = fs.rename(sourceFile, archiveFile) &&
+            fs.delete(sourceFile.getParent, true) &&
+            fs.setPermission(archiveFile, new FsPermission(archivePerms)).equals() // "equals" is here to get a bool return.
+        if (result) {
+            log.info(s"Archive created: ${archiveFile.toString}")
+        }
+        result
+    }
+}
```
Comment (jobar) [inline src/main/scala/org/wikimedia/analytics/hdfstools/HDFSArchiver.scala]: There should be a newline at the end of the file

**Review by jobar — COMMENTED** [review]
First review, code only - will be testable after today's refinery-source deployment

File: pom.xml (hunk 1 above)
Comment (ebysans) [reply pom.xml]: Changed in the next commit

File: src/test/scala/org/wikimedia/analytics/hdfstools/TestHDFSArchiver.scala (hunk 2 above)
Comment (ebysans) [reply src/test/scala/org/wikimedia/analytics/hdfstools/TestHDFSArchiver.scala]: Noted

File: src/main/scala/org/wikimedia/analytics/hdfstools/HDFSArchiver.scala (hunk 3 above)
Comment (ebysans) [reply src/main/scala/org/wikimedia/analytics/hdfstools/HDFSArchiver.scala]: Got it. Will add in the next commit

File: pom.xml [hunk 4]
```diff
@@ -116,15 +116,21 @@
 
         <dependency>
             <groupId>org.scalatest</groupId>
-            <artifactId>scalatest_2.11</artifactId>
-            <version>3.1.0</version>
+            <artifactId>scalatest_2.12</artifactId>
+            <version>3.0.5</version>
```
Comment (jobar) [inline pom.xml]: Let's keep the version we were using (it was 3.1.0) even if we change the scala version.

File: changelog.md [hunk 5]
```diff
@@ -1,3 +1,6 @@
+# 0.0.6
+ * Add HDFSArchiver, a hdfs tool.
```
Comment (jobar) [inline changelog.md]: Could you please add a line about fixing the hdfs-rsync tests and correcting a bug in update mode? Many thanks :)

File: pom.xml [hunk 6]
```diff
@@ -116,15 +116,21 @@
...
             <scope>test</scope>
         </dependency>
 
         <dependency>
             <groupId>com.github.scopt</groupId>
-            <artifactId>scopt_2.11</artifactId>
-            <version>3.7.1</version>
+            <artifactId>scopt_2.12</artifactId>
+            <version>3.5.0</version>
```
Comment (jobar) [inline pom.xml]: Same - let's keep the version originally used (3.7.1)

File: src/test/scala/org/wikimedia/analytics/hdfstools/TestHdfsRsyncHelper.scala [hunk 7]
```diff
@@ -7,11 +7,10 @@ import org.apache.commons.io.FileUtils
 import org.apache.log4j.spi.LoggingEvent
 import org.apache.log4j.{AppenderSkeleton}
 import org.scalatest.BeforeAndAfterEach
-import org.scalatest.flatspec.AnyFlatSpec
-import org.scalatest.matchers.should.Matchers
+import org.scalatest.{FlatSpec, Matchers}
 
 
-trait TestHdfsRsyncHelper extends AnyFlatSpec with Matchers with BeforeAndAfterEach {
+trait TestHdfsRsyncHelper extends FlatSpec with Matchers with BeforeAndAfterEach {
```
Comment (jobar) [inline src/test/scala/org/wikimedia/analytics/hdfstools/TestHdfsRsyncHelper.scala]: If we don't change the version of scalatest we can keep this to `AnyFlatSpec

**Review by jobar — COMMENTED** [review]
Some small changes requested - then merge!

File: pom.xml (hunk 4 above)
Comment (ebysans) [reply pom.xml]: The build fails when I using the older scalatest version and scopt version. Hence the reason I changed to the earlier version to be compatible with scala 2.12

Caused by: java.lang.NoClassDefFoundError: scala/Product$class
        at org.scalatest.Outcome.<init>(Outcome.scala:47)
        at org.scalatest.Succeeded$.<init>(Outcome.scala:357)
        at org.scalatest.Succeeded$.<clinit>(Outcome.scala)
        at org.scalatest.Assertions$class.$init$(Assertions.scala:1186)
        at org.scalatest.flatspec.AnyFlatSpec.<init>(AnyFlatSpec.scala:1685)
        at org.wikimedia.analytics.hdfstools.TestHdfsRsyncExec.<init>(TestHdfsRsyncExec.scala:13)
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
        at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:128)
        at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:341)
        at java.base/java.lang.Class.newInstance(Class.java:677)
        at org.apache.maven.surefire.junit.PojoTestSet.<init>(PojoTestSet.java:63)
        at org.apache.maven.surefire.junit.JUnit3Provider.createTestSet(JUnit3Provider.java:107)
        at org.apache.maven.surefire.junit.JUnit3Provider.invoke(JUnit3Provider.java:97)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        ... 6 more

**Review by jobar — APPROVED** [review]
LGTM ! Thanks @ebysans :)

File: pom.xml (hunk 4 above)
Comment (ebysans) [reply pom.xml]: Done

File: changelog.md (hunk 5 above)
Comment (ebysans) [reply changelog.md]: This has been added.

File: pom.xml (hunk 6 above)
Comment (ebysans) [reply pom.xml]: Noted. will be done in next commit.

File: src/test/scala/org/wikimedia/analytics/hdfstools/TestHdfsRsyncHelper.scala (hunk 7 above)
Comment (ebysans) [reply src/test/scala/org/wikimedia/analytics/hdfstools/TestHdfsRsyncHelper.scala]: Thanks for this. New commit will have this change

**Pull request merged**